### PR TITLE
Add oauth providers via cli

### PR DESCRIPTION
**Pull request by techknowlogick**
Related to https://github.com/go-gitea/gitea/issues/183#issuecomment-261824704

cc: @lunny 

TODO:
- [x] Documentation

### CODE REVIEW

**Comment by lunny** [comment]
@techknowlogick but maybe you can consider `add`, `delete`, `update` subcommands?

File: cmd/admin.go [hunk 1]
```diff
@@ -121,6 +123,82 @@ var (
...
+			},
+			cli.StringFlag{
+				Name:  "use-custom-urls",
+				Value: "false",
+				Usage: "Use custom URLs for GitLab/GitHub OAuth endpoints",
+			},
+			cli.StringFlag{
+				Name:  "custom-auth-url",
+				Value: "",
+				Usage: "Use a custom Authorize URL (option for GitLab/GitHub)",
```
Comment (adelowo) [inline cmd/admin.go]: Shouldn't this be `Use a custom Authorization URL` ?

Comment (adelowo) [inline cmd/admin.go]: Shouldn't this be `Use a custom Authorization URL` ?

**Comment by techknowlogick** [comment]
@lunny I've added `delete`, `update` and `list` commands.

File: cmd/admin.go [hunk 2]
```diff
@@ -121,6 +125,177 @@ var (
...
+				Usage: "ID of authentication source that will be deleted",
+			},
+		},
+	}
+
+	microcmdAuthUpdateOauth = cli.Command{
+		Name:   "update-oauth",
+		Usage:  "Update existing Oauth authentication source",
+		Action: runUpdateOauth,
+		Flags: []cli.Flag{
```
Comment (lafriks) [inline cmd/admin.go]: Maybe this flag array could be moved out to array so that it could be reused in both add and update?

Comment (techknowlogick) [reply cmd/admin.go]: thanks @lafriks updated.

**Comment by techknowlogick** [comment]
@lafriks changes made. Please review.

File: cmd/admin.go [hunk 3]
```diff
@@ -121,6 +125,123 @@ var (
...
+	}
+
+	oauthCLIFlags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "config, c",
+			Value: "custom/conf/app.ini",
+			Usage: "Custom configuration file path",
+		},
+		cli.Int64Flag{
+			Name:  "id",
```
Comment (lafriks) [inline cmd/admin.go]: You could insert it id in this position only for add action

Comment (techknowlogick) [reply cmd/admin.go]: updated

File: cmd/admin.go [hunk 4]
```diff
@@ -121,6 +125,121 @@ var (
...
+			Value: "",
+			Usage: "Use a custom Email URL (option for GitHub)",
+		},
+	}
+
+	microcmdAuthUpdateOauth = cli.Command{
+		Name:   "update-oauth",
+		Usage:  "Update existing Oauth authentication source",
+		Action: runUpdateOauth,
+		Flags:  append(oauthCLIFlags, idFlag),
```
Comment (lafriks) [inline cmd/admin.go]: maybe `append(oauthCLIFlags[:1], append([]cli.Flag{idFlag}, oauthCLIFlags[1:]...)...)`? To have better order :)

**Review by lafriks — APPROVED** [review]

**Review by Bwko — APPROVED** [review]

**Pull request merged**